langserver/fs.go

```diff
@@ -26,42 +26,59 @@ func IsFileSystemRequest(method string) bool {
 		method == "textDocument/didSave"
 }
 
-func (h *HandlerShared) HandleFileSystemRequest(ctx context.Context, req *jsonrpc2.Request) error {
+// HandleFileSystemRequest handles textDocument/did* requests. The path the
+// request is for is returned. true is returned if a file was modified.
+func (h *HandlerShared) HandleFileSystemRequest(ctx context.Context, req *jsonrpc2.Request) (string, bool, error) {
 	span := opentracing.SpanFromContext(ctx)
 	h.Mu.Lock()
 	overlay := h.overlay
 	h.Mu.Unlock()
 
+	var before []byte
+	checkChanged := func(uri string) bool {
+		// We ignore readFile errors, since the []byte being nil gives
+		// us the same comparison.
+		b, _ := h.readFile(ctx, uri)
+		return !bytes.Equal(before, b)
+	}
+
 	switch req.Method {
 	case "textDocument/didOpen":
 		var params lsp.DidOpenTextDocumentParams
 		if err := json.Unmarshal(*req.Params, &params); err != nil {
-			return err
+			return "", false, err
 		}
-		span.SetTag("uri", params.TextDocument.URI)
+		uri := params.TextDocument.URI
+		span.SetTag("uri", uri)
+		before, _ = h.readFile(ctx, uri)
```

Comment:
I was going from the assumption that an error means file not found => empty []byte is file not found, which makes the bytes.Equal check I do correct. However, there are (rarer) cases were we can get an error for a different reason. I'll correctly handle that.